sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ResourceEvaluationStage.java

```diff
@@ -0,0 +1,189 @@
+package com.mesosphere.sdk.offer.evaluate;
+
+import com.google.protobuf.TextFormat;
+import com.mesosphere.sdk.offer.*;
+import org.apache.mesos.Protos;
+import org.apache.mesos.Protos.Resource;
+import org.apache.mesos.Protos.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * This class evaluates an offer against a given {@link OfferRequirement}, ensuring that it contains a sufficient amount
+ * or value of the supplied {@link Resource}, and creating a {@link ReserveOfferRecommendation} or
+ * {@link UnreserveOfferRecommendation} where necessary.
+ */
+public class ResourceEvaluationStage implements OfferEvaluationStage {
+    private static final Logger logger = LoggerFactory.getLogger(OfferEvaluator.class);
+
+    private ResourceRequirement resourceRequirement;
+    private Optional<String> taskName;
```

Comment:
I think the best practice for Optionals is that they aren't fields.  I know we have broken this rule a lot.  Functions should return Optionals as is done in the `getTaskName` method.